coordinator/src/main/scala/filodb.coordinator/queryplanner/SingleClusterPlanner.scala

```diff
@@ -269,45 +304,143 @@ class SingleClusterPlanner(val dataset: Dataset,
+      case lp: TopLevelSubquery            => materializeTopLevelSubquery(qContext, lp, forceInProcess)
+      case lp: LabelCardinality            => materializeLabelCardinality(qContext, lp, forceInProcess)
       //case _                               => throw new BadQueryException("Invalid logical plan")
     }
   }
   // scalastyle:on cyclomatic.complexity
 
-  override def materializeBinaryJoin(qContext: QueryContext,
-                                     lp: BinaryJoin): PlanResult = {
-    val lhs = walkLogicalPlanTree(lp.lhs, qContext)
+  /**
+   * Returns the shards spanned by a LogicalPlan's data.
+   * @return an occupied Option iff shards could be determined for all leaf-level plans.
+   */
+  private def getShardSpanFromLp(lp: LogicalPlan,
+                                 qContext: QueryContext): Option[Set[Int]] = {
+
+    case class LeafInfo(lp: LogicalPlan,
+                        filters: Set[ColumnFilter],
+                        tslabels: Option[Seq[String]])
+
+    // construct a LeafInfo for each leaf
+    val leaves = findLeafLogicalPlans(lp).map{ leafPlan =>
+      val filters = getColumnFilterGroup(leafPlan)
+      assert(filters.size == 1, s"expected leaf plan to yield single filter group, but got ${filters.size}")
+      val targetSchemaProvider = qContext.plannerParams.targetSchemaProvider
+      if (targetSchemaProvider.isDefined) {
+        val tsFunc = targetSchemaProvider.get.targetSchemaFunc(filters.head.toSeq)
+        val (start, stop) = leafPlan match {
+          // TODO(a_theimer): other leaf plan types? Range selector types?
+          case rs: RawSeries => rs.rangeSelector match {
+            case is: IntervalSelector => (is.from, is.to)
+            case _ => throw new IllegalArgumentException(s"unhandled type: ${rs.rangeSelector.getClass}")
+          }
+          case _ => throw new IllegalArgumentException(s"unhandled type: ${leafPlan.getClass}")
+        }
+        val tsLabels = findTargetSchema(tsFunc, start, stop).map(_.schema)
+        LeafInfo(leafPlan, filters.head, tsLabels)
+      } else {
+        LeafInfo(leafPlan, filters.head, None)
+      }
+    }
+
+    // if we can't extract shards from all filters, return an empty result
+    if (!leaves.forall{ leaf =>
+      canGetShardsFromFilters(leaf.filters.toSeq, qContext)
+    }) return None
+
+    Some(leaves.flatMap{ leaf =>
+      val colFilterLabels = leaf.filters.map(_.column)
+      val useTargetSchema = leaf.tslabels.isDefined &&
+        leaf.tslabels.get.toSet.subsetOf(colFilterLabels)
+      shardsFromFilters(leaf.filters.toSeq, qContext, useTargetSchema)
+    }.toSet)
+  }
+
+  // scalastyle:off method.length
+  // scalastyle:off cyclomatic.complexity
+  /**
+   * Returns the set of shards to which this BinaryJoin can be pushed down.
+   * @return an occupied Option iff it is valid to perform the a pushdown optimization on this BinaryJoin.
+   */
+  private def getBinaryJoinPushdownShards(qContext: QueryContext,
```

Comment:
If we end up adding aggregator pushdowns, this method is pretty easily converted to a general "getPushdownShards" for either pushdown optimization.